Introducing C++ Swift Interop

nmahendru opened the pull request:
#### e88f01ab1a57ebb1b114ee93abd0867a6ec0cac8
<pre>
Introducing C++ Swift Interop
<a href="https://bugs.webkit.org/show_bug.cgi?id=271263">https://bugs.webkit.org/show_bug.cgi?id=271263</a>
<a href="https://rdar.apple.com/123331949">rdar://123331949</a>

Reviewed by Alex Christensen, Geoffrey Garen and Elliott Williams.

This is an examplar of the C++ Swift interop feature.
We have randomly chosen the Digest Implementation to be the first one to adopt this.
The Compiler feature is only available for newer platforms to code has been guarded
with ifdefs.

The changes in AbortWithReasonSPI.h are to allow the Swift Generator to work.
It, at the moment cannot handle headers included inside `extern C` blocks.

Input parameteres to Swift have been intentionally chosen as spans as
swift, at the moment my make copies of inputs for safety guarantees.
Copying a span is cheap so that is Okie.
On the return path, Swift calls the WTF::Vector move constructor(confirmed in debugger).

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesGcm.test(_:)):
(AesKwRV.errCode):
(AesKwRV.outputSize):
(HashFunction.update(_:)):
(Digest.sha1(_:)):
(Digest.sha256(_:)):
(Digest.sha384(_:)):
(Digest.sha512(_:)):
(AesKwReturnValue.errCode): Deleted.
(AesKwReturnValue.outputSize): Deleted.
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift: Added.
(HashFunction.update(_:)):
* Source/WebCore/PAL/pal/PALSwiftModule.h: Copied from Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h.
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp: Renamed from Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm.
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/crypto/openssl/CryptoDigestOpenSSL.cpp:
(PAL::createCryptoDigest):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/module.modulemap: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::dispatchDigest):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.mm.
(WebCore::encryptCryptoKitAESGCM):
(WebCore::CryptoAlgorithmAESGCM::platformEncrypt):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm.
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoUtilitiesCocoa.cpp:

Canonical link: <a href="https://commits.webkit.org/276651@main">https://commits.webkit.org/276651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be30731bbd0b5ff495f830604f285f1498675df0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45261 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47781 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47924 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41269 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47568 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28570 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21776 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21428 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/39039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18220 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18840 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40119 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3308 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/38479 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41524 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40436 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49629 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/44729 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/20243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16771 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21549 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42959 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51889 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6295 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21237 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/10582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -188,108 +189,45 @@ public final class AesKw: NSObject {
-            data: UnsafeBufferPointer(start: data, count: Int(dataSize)), out: &d, SHA384.self)
+extension PALBytes: ContiguousBytes {
+    borrowing public func withUnsafeBytes<R>(_ body: (UnsafeRawBufferPointer) throws -> R) rethrows -> R {
+        try body(UnsafeRawBufferPointer(start:self.__dataUnsafe(), count: self.size()))
```
Do we have any options to reduce our use of unsafety here? Or do we at least understand what safety might look like in the future?

nmahendru replied:
The underlying CryptoKitAPI eventually needs an UnsafeBufferPointer. Now matter how we wrap it we will have an unsafe somewhere, if not at this level then inside the wrapper around the API that expects UnsafeBufferPointer.

achristensen07 commented on Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h:
```diff
@@ -34,8 +34,7 @@ extern "C" {
     || PLATFORM(VISION)
-
-#if USE(APPLE_INTERNAL_SDK)
+#if USE(APPLE_INTERNAL_SDK) && !HAVE(SWIFT_CPP_INTEROP)
```
I don't think we should do this always on all systems that have swift/cpp interop.  I think we should do this only when including this header as part of swift/cpp interop.  That way, we still use the contents of sys/reason.h on new systems.

achristensen07 commented on Source/WebCore/PAL/pal/crypto/openssl/CryptoDigestOpenSSL.cpp:
```diff
@@ -113,6 +113,7 @@ CryptoDigest::~CryptoDigest()
 
 static std::unique_ptr<CryptoDigestContext> createCryptoDigest(CryptoDigest::Algorithm algorithm)
 {
+    std::unique_ptr<CryptoDigest> digest(new CryptoDigest);
```
Can we use makeUnique instead of new?
Also, this seems unused.

nmahendru replied:
yea this is rebase remnant!! removed.

achristensen07 commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp:
```diff
@@ -26,9 +26,13 @@
 #include "config.h"
 #include "CryptoAlgorithmAESKW.h"
 
+
```
unneeded extra space.

nmahendru replied:
removed

achristensen07 commented on Source/WebCore/crypto/cocoa/CryptoUtilitiesCocoa.cpp:
```diff
@@ -28,6 +28,7 @@
 
 #include "CryptoAlgorithmAESCTR.h"
 #include <CommonCrypto/CommonCrypto.h>
+#include <pal/spi/cocoa/CommonCryptoSPI.h>
```
Is this needed?

nmahendru replied:
yes this was a compiler error if I remember correctly.

achristensen07 commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
```diff
@@ -81,10 +86,15 @@
     return WTFMove(plainText);
 }
 
+#if HAVE(SWIFT_CPP_INTEROP)
```
I think it would be a lot nicer for our { and } matching to have this #if inside the function, then just have an #else UNUSED_PARAM(useCryptoKit).  Same everywhere.

nmahendru replied:
done

achristensen07 commented on Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h:
```diff
@@ -51,7 +50,7 @@ void abort_with_reason(uint32_t reason_namespace, uint64_t reason_code, const ch
 #endif
 
-#if !USE(APPLE_INTERNAL_SDK)
+#if !USE(APPLE_INTERNAL_SDK) || (USE(APPLE_INTERNAL_SDK) && defined(SWIFT_ONLY))
```
The second USE(APPLE_INTERNAL_SDK) isn't needed.
I'd call SWIFT_ONLY something like WK_COMPILING_SWIFT

emw-apple replied:
I think you can use `__swift__` to check whether the header is being read by swift's clang importer. So you don't need to define your own SWIFT_ONLY macro.

nmahendru replied:
renamed to WK_SWIFT_INTEROP_HEADER_GENERATION

nmahendru replied:
> I think you can use `__swift__` to check whether the header is being read by swift's clang importer. So you don't need to define your own SWIFT_ONLY macro.

Will this always be true and never break in future ?

emw-apple replied:
I mean, it's part of how the clang importer works and is used extensively in Swift's standard library: https://github.com/search?q=repo%3Aapple%2Fswift%20__swift__&type=code

So, yes, I think it is as stable as other compiler-provided macros like `__objc__`.

achristensen07 commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -43620,6 +43621,26 @@
+			);
+			inputPaths = (
+			);
+			name = "Remove stale PALSwift.h added in 275246@main";
```
This is temporary, right?

nmahendru replied:
yes temporary, so that everyone does not have to delete the file that they will have from main.

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -53,7 +53,8 @@ FRAMEWORK_SEARCH_PATHS[sdk=macosx*] = $(WK_QUOTED_OVERRIDE_FRAMEWORKS_DIR);
 SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks $(SDK_DIR)$(SYSTEM_LIBRARY_DIR)/Frameworks;
 
-HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore $(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH) $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(OBJROOT)/PAL.build/PAL.build/DerivedSources $(OBJROOT)/PAL.build/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)/PA ...
+HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore $(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH) $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(OBJROOT)/PAL.build/PAL.build/DerivedSources $(OBJROOT)/PAL.build/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)/PA ...
```
> $(SRCROOT)/PAL/pal;

Why do we need to do this? The way PAL headers have always worked is they are marked "Private" and get copied to usr/local/include/pal in the build products directory. Clients import them via angle brackets like `#include <pal/...>`.

nmahendru replied:
If I do `<pal/PALModule.h>` followed by `<PALSwift.h>` the style checker complains about sorting.
So this is my fix for that.
and the ordering here is important and I would like to keep the names.

emw-apple replied:
Changing search path logic to satisfy the style checker seems like incorrect prioritization to me. Let me see if I can come up with an alternative, it's pretty uncommon for us to have headers with strict include order requirements…

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -67,5 +67,55 @@ DAV1D_LIBTOOLFLAGS_YES = $(BUILT_PRODUCTS_DIR)/libdav1d.a;
+SWIFT_OBJC_INTEROP_MODE_MACCATALYST_SINCE_14 = objcxx;
+SWIFT_OBJC_INTEROP_MODE_IOS_SINCE_17 = objcxx;
+SWIFT_OBJC_INTEROP_MODE_WATCHOS_SINCE_10 = objcxx;
+SWIFT_OBJC_INTEROP_MODE_TVOS_SINCE_17 = objcxx;
```
Do we actually need to _avoid_ setting this on older platforms? If we aren't compiling swift sources on pre-2023 OS versions, then I would expect that this setting to have no effect. So we could just set it unconditionally.

nmahendru replied:
The generator code does trigger on every platform(eg Monterey) so I am just locking this down as much as possible. But maybe now as I have `SWIFTCPP_INTEROP` this probably can be set unconditionally

nmahendru replied:
done.

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -67,5 +67,55 @@ DAV1D_LIBTOOLFLAGS_YES = $(BUILT_PRODUCTS_DIR)/libdav1d.a;
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_MACCATALYST_SINCE_14 = SWIFTCPP_INTEROP;
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_IOS_SINCE_17 = SWIFTCPP_INTEROP;
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_WATCHOS_SINCE_10 = SWIFTCPP_INTEROP;
+SWIFT_ACTIVE_COMPILATION_CONDITIONS_TVOS_SINCE_17 = SWIFTCPP_INTEROP;
```
What is this for? I don't see a `SWIFTCPP_INTEROP` condition checked anywhere.

nmahendru replied:
see `CryptoKitShim.swift`

emw-apple replied:
My bad, large diff = hidden by default. Thanks!

emw-apple commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -23,273 +23,220 @@
-    case decryptionFailed = 6
-    case hashingFailed = 7
-}
+    #if SWIFTCPP_INTEROP
```
Do we actually need this SWIFTCPP_INTEROP check? Would we ever try to compile this code when interop is not available?

nmahendru replied:
The swift generator code is triggered for instance on Monterey so I am just guarding the whole file.

nmahendru replied:
also I am being explicit about it honestly. `swift(>=5.9)` might just be enough but that is not really informative.

emw-apple commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -23,273 +23,220 @@
-}
+    #if SWIFTCPP_INTEROP
+        import CryptoKit
+        import Foundation
```
Formatting nit: WebKit's style is to not have preprocessor directives (`#if`, etc.) add a level of indentation to the code. I see no reason why that should not apply to Swift code, too. So these `#if`s should have no leading whitespace and the code inside them should be normally indented.

nmahendru replied:
I am using the `Tools/Script/swift/format` for format the code.
Let me try to see if that can be modified to achieve this.

nmahendru replied:
done.

emw-apple replied:
Yeah, my understanding is that swift-format is using built-in rules, not necessarily WebKit style? I wouldn't try too hard to make it match our style, as until and unless it becomes part of `check-webkit-style` that script is only going to be run manually.

emw-apple commented on Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp:
```diff
@@ -189,35 +193,27 @@
-            if (WebCryptoErrorCodesSuccess != [WebCryptoDigest sha1:data.data() dataSize:data.size() digest:digest.data() digestSize:digest.size()])
-                return std::nullopt;
-            break;
+            return PAL::Digest::sha1(data);
```
nice!

emw-apple commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -43620,6 +43621,26 @@
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "rm -fv ${BUILT_PRODUCTS_DIR}/DerivedSources/WebCore/PALSwift.h\ntouch ${SCRIPT_OUTPUT_FILE_0}\n";
```
nit: If for some reason deletion fails, don't create the timestamp file.
```suggestion
			shellScript = "rm -fv ${BUILT_PRODUCTS_DIR}/DerivedSources/WebCore/PALSwift.h && touch ${SCRIPT_OUTPUT_FILE_0}\n";
```

nmahendru replied:
gotcha

emw-apple commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -43620,6 +43621,26 @@
 		};
+		947D341A2BA12DD50051D6D0 /* Remove stale PALSwift.h added in 275246@main */ = {
+			isa = PBXShellScriptBuildPhase;
+			alwaysOutOfDate = 1;
```
Now that you have an output edge, this should be removed so that the build system tracks it appropriately.
```suggestion
			alwaysOutOfDate = 0;
```

Will remove the message in build logs like:
> Run script build phase 'Remove stale PALSwift.h added in 275246@main' will be run during every build because the option to run the script phase "Based on dependency analysis" is unchecked.

nmahendru replied:
done.

emw-apple commented on Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h:
Commit message and code doesn't say why you're changing these, can you please explain why Swift isn't allow to include sys/reason.h? This is weird to me since it's a system header and surely other Swift projects import this.

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -53,7 +53,8 @@ FRAMEWORK_SEARCH_PATHS[sdk=macosx*] = $(WK_QUOTED_OVERRIDE_FRAMEWORKS_DIR);
 SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks $(SDK_DIR)$(SYSTEM_LIBRARY_DIR)/Frameworks;
 
-HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore $(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH) $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(OBJROOT)/PAL.build/PAL.build/DerivedSources $(OBJROOT)/PAL.build/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)/PA ...
+HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore $(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH) $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(OBJROOT)/PAL.build/PAL.build/DerivedSources $(OBJROOT)/PAL.build/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)/PA ...
```
Are you sure this is needed? For one, `$(SRCROOT)/PAL` is already on the header search paths. These are relative to SRCROOT, so the first "PAL" should be equivalent.

nmahendru replied:
I think I added this to overcome exactly the header not found problem. Let me try removing and give the compiler a chance to complain and if not I can remove it.

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -53,7 +53,8 @@ FRAMEWORK_SEARCH_PATHS[sdk=macosx*] = $(WK_QUOTED_OVERRIDE_FRAMEWORKS_DIR);
 
-HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore $(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH) $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(OBJROOT)/PAL.build/PAL.build/DerivedSources $(OBJROOT)/PAL.build/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)/PA ...
+HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore $(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH) $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(OBJROOT)/PAL.build/PAL.build/DerivedSources $(OBJROOT)/PAL.build/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)/PA ...
+
```
nit: unnecessary whitespace. HEADER_SEARCH_PATHS and SYSTEM_HEADER_SEARCH_PATHS so I think they should stay together.

nmahendru replied:
removed.

emw-apple replied:
Did you remove the line above (L55)? I was asking about below (L57) :P

Basically, this file should have no changes to it.

emw-apple commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -23,10 +23,14 @@
 
+#if swift(>=5.9)
+
+#if SWIFTCPP_INTEROP
```
SWIFTCPP_INTEROP is defined on 2023 platforms (macOS 14, iOS 17, etc.). Xcode 15 is aligned with those releases, and that's the Xcode that shipped Swift 5.9. So, are you sure SWIFTCPP_INTEROP is needed? I think we might be able to remove it, and the associated lines in PAL.xcconfig.

nmahendru replied:
SWIFTCPP_INTEROP is the flag that is added here to gate any swift code explicitly for platforms we don't want to support.
It's a little tricky with swift that even if you have `#if` still the compiler will parse the whole file, just not compile it. The file has some syntax that is not pre swift 5.9 friendly that's why swift 5.9 is added `outside` the SWIFTCPP_INTEROP.
`#if swift(>=5.9)` gets respected and the compiler does not parse. further if that is not true.
It is probably true that `swift(>=5.9)` is enough technically, but I am intentionally being explicit about it.

emw-apple replied:
Sure, but it's just another conditional and bit of complexity. And since it's not based on the Platform.h flag, it's liable to get out of sync from HAVE_SWIFT_CPP_INTEROP and surprise someone later. I'm not saying you need to figure out a way to bridge Platform.h flags into Swift (though someone should!) but I'm not sure the redeclaration adds much value.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -0,0 +1,35 @@
+#include <cstdint>
+#include <wtf/Vector.h>
+
+using PALBytes = WTF::Vector<uint8_t>;
```
Let's call this something like VectorUInt8. We want to be clear that all we've done is write out the template specialization for Vector<uint8_t>, and we want to avoid naming conflict with future specializations, which will be required because Swift only understands specializations and not templates.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -66,7 +69,8 @@ public final class AesGcm: NSObject {
                 return .invalidArgument
             }
-            let nonce = try AES.GCM.Nonce(data: UnsafeBufferPointer(start: iv, count: Int(ivSize)))
+            let nonce = try AES.GCM.Nonce(
+                data: UnsafeBufferPointer(start: iv, count: Int(ivSize)))
```
Our goal is to write Swift code that is at least consistent with some future vision for safety, even if Swift is not yet safe. If Nonce() is expressed as an unsafe pointer API, I think we want an explicit wrapper or extension or replacement for it, so that our code is expressed in terms of safety. Maybe we need a separate file called "CryptoKitUnsafeAPIWrappers" or something like that, to hold all these abstractions.

nmahendru replied:
This PR is only for Digest. I am going to cleanup the AESGCM, AESKW wrappers in a future PRs.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -92,31 +96,30 @@ public final class AesGcm: NSObject {
             }
             sealedBox.ciphertext.copyBytes(to: cipherText, count: Int(plainTextSize))
-            sealedBox.tag.copyBytes(to: cipherText + Int(plainTextSize), count: sealedBox.tag.count)
+            sealedBox.tag.copyBytes(
+                to: cipherText + Int(plainTextSize), count: sealedBox.tag.count)
```
No bounds check here, so we need a wrapper that expresses the bounds check.

nmahendru replied:
https://github.com/WebKit/WebKit/pull/25602#discussion_r1531120724

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -131,7 +134,8 @@ public final class AesKw: NSObject {
             let cipher = try AES.KeyWrap.wrap(
                 SymmetricKey(data: UnsafeBufferPointer(start: data, count: Int(dataSize))),
-                using: SymmetricKey(data: UnsafeBufferPointer(start: key, count: Int(keySize))))
+                using: SymmetricKey(
+                    data: UnsafeBufferPointer(start: key, count: Int(keySize))))
```
Need a wrapper that expresses a safe way to construct a key.

geoffreygaren replied:
There are other cases like this too, but I won't comment on each one. One way to get a close approximation is to search for "unsafe" in the code; though some unsafe pointers will instantiate implicitly.

nmahendru replied:
https://github.com/WebKit/WebKit/pull/25602#discussion_r1531120724

emw-apple commented on Source/WebCore/PAL/pal/crypto/openssl/CryptoDigestOpenSSL.cpp:
```diff
@@ -154,5 +154,4 @@ std::optional<Vector<uint8_t>> CryptoDigest::computeHash(CryptoDigest::Algorithm
     digest->m_context->addBytes(data.data(), data.size());
     return digest->m_context->computeHash();
 }
-
```
nit: probably don't need to delete this whitespace line, especially since you are not making any code changes to this file.

nmahendru replied:
removed.

emw-apple reviewed (approved):
r+ with formatting nits

cdumez commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -28,7 +28,7 @@
 #include <wtf/Vector.h>
 
-using VectorUInt8 = WTF::Vector<uint8_t>;
+using VectorUInt8 = WTF::Vector<uint8_t, 0, CrashOnOverflow, 16, WTF::VectorBufferMalloc, WTF::IsCopyable::No>;
```
Can't we just pass std::spans to Swift? Why do we need to pass vectors to these functions?

cdumez replied:
If we'd pass a `std::span<const uint8_t>`, we wouldn't care if Swift copies it since it is cheap to copy

nmahendru replied:
Removed the nonCopyable and using spans instead.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -0,0 +1,36 @@
+#include <wtf/Vector.h>
+
+using VectorUInt8 = WTF::Vector<uint8_t>;
+using SpanUInt8 = std::span<const uint8_t>;
```
Let's call this SpanConstUInt8, since const is a part of the type.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -188,108 +193,42 @@ public final class AesKw: NSObject {
+        self.update(
+            bufferPointer: UnsafeRawBufferPointer(
+                start: data.__dataUnsafe(), count: data.size()))
+    }
```
To demonstrate our discipline of programming using safe constructs, and our expectation that future APIs will expose safe constructs, I think it's good for overlays that shim unsafe APIs like this to live in a separate file. Maybe "CryptoKitOverlay.swift", or we can put them all together in "UnsafeOverlays.swift".

If you'd like, you can make this change in the follow-up patch that removes the other unsafe pointers from this file.

geoffreygaren reviewed (approved):
r=me

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -188,108 +193,42 @@ public final class AesKw: NSObject {
+        hasher.update(data: data)
+        let digest = hasher.finalize()
+        var returnValue = VectorUInt8(T.Digest.byteCount)
+        returnValue.fill(0)
```
No need to fill with zeroes here.

geoffreygaren commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -188,108 +193,42 @@ public final class AesKw: NSObject {
+        let digest = hasher.finalize()
+        var returnValue = VectorUInt8(T.Digest.byteCount)
+        returnValue.fill(0)
+        var counter = 0
```
We usually call this variable "index".

nmahendru committed:
Introducing C++ Swift Interop
https://bugs.webkit.org/show_bug.cgi?id=271263
rdar://123331949

Reviewed by Alex Christensen, Geoffrey Garen and Elliott Williams.

This is an examplar of the C++ Swift interop feature.
We have randomly chosen the Digest Implementation to be the first one to adopt this.
The Compiler feature is only available for newer platforms to code has been guarded
with ifdefs.

The changes in AbortWithReasonSPI.h are to allow the Swift Generator to work.
It, at the moment cannot handle headers included inside `extern C` blocks.

Input parameteres to Swift have been intentionally chosen as spans as
swift, at the moment my make copies of inputs for safety guarantees.
Copying a span is cheap so that is Okie.
On the return path, Swift calls the WTF::Vector move constructor(confirmed in debugger).

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesGcm.test(_:)):
(AesKwRV.errCode):
(AesKwRV.outputSize):
(HashFunction.update(_:)):
(Digest.sha1(_:)):
(Digest.sha256(_:)):
(Digest.sha384(_:)):
(Digest.sha512(_:)):
(AesKwReturnValue.errCode): Deleted.
(AesKwReturnValue.outputSize): Deleted.
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift: Added.
(HashFunction.update(_:)):
* Source/WebCore/PAL/pal/PALSwiftModule.h: Copied from Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h.
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp: Renamed from Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm.
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/crypto/openssl/CryptoDigestOpenSSL.cpp:
(PAL::createCryptoDigest):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/module.modulemap: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::dispatchDigest):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.mm.
(WebCore::encryptCryptoKitAESGCM):
(WebCore::CryptoAlgorithmAESGCM::platformEncrypt):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm.
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoUtilitiesCocoa.cpp:

Canonical link: https://commits.webkit.org/276651@main
